quickjs.c

```diff
@@ -45531,180 +45531,180 @@ static int js_json_to_str(JSContext *ctx, JSONStringifyContext *jsc,
+                }
+
+                JSValue prop_val = JS_GetProperty(ctx, val, props[i].atom);
+                if (JS_IsException(prop_val))
                     goto exception;
-                v = JS_GetPropertyValue(ctx, val, js_dup(prop));
-                if (JS_IsException(v))
+
+                JSValue prop_str = JS_AtomToString(ctx, props[i].atom);
+                if (JS_IsException(prop_str)) {
+                    JS_FreeValue(ctx, prop_val);
                     goto exception;
-                v = js_json_check(ctx, jsc, val, v, prop);
-                if (JS_IsException(v))
+                }
+
+                if (string_buffer_putc8(jsc->b, '"') < 0 ||
+                    string_buffer_concat_value(jsc->b, prop_str) < 0 ||
+                    string_buffer_putc8(jsc->b, '"') < 0 ||
+                    string_buffer_putc8(jsc->b, ':') < 0) {
+                    JS_FreeValue(ctx, prop_val);
+                    JS_FreeValue(ctx, prop_str);
                     goto exception;
-                if (!JS_IsUndefined(v)) {
-                    if (has_content)
-                        string_buffer_putc8(jsc->b, ',');
-                    prop = JS_ToQuotedStringFree(ctx, prop);
-                    if (JS_IsException(prop)) {
-                        JS_FreeValue(ctx, v);
-                        goto exception;
-                    }
-                    string_buffer_concat_value(jsc->b, sep);
-                    string_buffer_concat_value(jsc->b, prop);
-                    string_buffer_putc8(jsc->b, ':');
-                    string_buffer_concat_value(jsc->b, sep1);
-                    if (js_json_to_str(ctx, jsc, val, v, indent1))
-                        goto exception;
-                    has_content = true;
                 }
+
+                JS_FreeValue(ctx, prop_str);
+
+                if (js_json_to_str(ctx, jsc, val, prop_val, indent1) < 0)
+                    goto exception;
             }
-            if (has_content && JS_VALUE_GET_STRING(jsc->gap)->len != 0) {
-                string_buffer_putc8(jsc->b, '\n');
-                string_buffer_concat_value(jsc->b, indent);
-            }
-            string_buffer_putc8(jsc->b, '}');
+
+            if (string_buffer_putc8(jsc->b, '}') < 0)
+                goto exception;
+
+            js_free_prop_enum(ctx, props, prop_count);
         }
-        if (check_exception_free(ctx, js_array_pop(ctx, jsc->stack, 0, NULL, 0)))
-            goto exception;
-        JS_FreeValue(ctx, val);
-        JS_FreeValue(ctx, tab);
+
+        JS_FreeValue(ctx, indent1);
         JS_FreeValue(ctx, sep);
         JS_FreeValue(ctx, sep1);
-        JS_FreeValue(ctx, indent1);
+        JS_FreeValue(ctx, tab);
         JS_FreeValue(ctx, prop);
         return 0;
+
+exception:
+        JS_FreeValue(ctx, indent1);
+        JS_FreeValue(ctx, sep);
+        JS_FreeValue(ctx, sep1);
+        JS_FreeValue(ctx, tab);
+        JS_FreeValue(ctx, prop);
```

Comment:
Move the definitions to the outer scope and add `js_free_prop_enum(ctx, props, prop_count);`